### ISSUE TITLE
URSACHE: 3500 (IllegalStateException) - exception during risk level calculation

### ISSUE THREAD

**Issue by tvannahl** [corona-warn-app/cwa-app-android#603]
## Describe the bug

Error message popped up after opening the app aproximately one day after installation. The error can be triggered again when clicking on "Risk factor unknown".

Bluetooth hasn't been active for the whole duration since installation, and a few trips have been made with active bluetooth.

## Expected behaviour

No error should occur

## Steps to reproduce the issue

Unknown

## Technical details

- Host Machine OS (Windows/Linux/Mac): Huawei Nova 2 (Android 8.0.0)

## Additional context

![signal-2020-06-17-170844_001](https://user-images.githubusercontent.com/662132/84917587-d068e980-b0bf-11ea-8003-d2945378c30e.jpeg)
![signal-2020-06-17-170844_002](https://user-images.githubusercontent.com/662132/84917621-da8ae800-b0bf-11ea-9684-7f60ab762cfe.jpeg)

**Comment by pwoessner** [comment]
Hi @tvannahl , first of all: if you deactivate bluetooth the whole time the tracing won't work as indicated by the app. Your trips will then not record any contacts. 

Nevertheless, this error should not appear and therefore we have some follow up questions: 

- have you activated tracing in the onboarding and then switched off bluetooth later?
- do you have a timestamp on the main screen when the app was updated the last time? It is on the Risk Card with the label: "Aktualisiert: " or "Updated: "? 

Thanks, Philipp

**Comment by tvannahl** [comment]
> Hi @tvannahl , first of all: if you deactivate bluetooth the whole time the tracing won't work as indicated by the app. Your trips will then not record any contacts.

Thank you for the quick response. We are fully aware of this, the bluetooth was just deactivated whilst being in door, in order to save battery and has been activated when leaving the house in order to make full use of the app.

>     * have you activated tracing in the onboarding and then switched off bluetooth later?

The tracing has been activated during the onboarding, the user is not entirely sure if the bluetooth has been deactivated directly after the first installation or with a bigger delay.

>     * do you have a timestamp on the main screen when the app was updated the last time? It is on the Risk Card with the label: "Aktualisiert: " or "Updated: "?

No there is no such timestamp, only "Unknown risk" and "updated daily".

**Comment by pwoessner** [comment]
Ok thanks for your answers. Can you also add a photo of the risk card with the detailed text below the "Unbekanntes Risiko"? Or if a photo is too complicated just summarise what the text is saying?

**Comment by tvannahl** [comment]
Sure thing:

![signal-2020-06-17-180550](https://user-images.githubusercontent.com/662132/84922030-93075a80-b0c5-11ea-8aa9-7ccb7d47916d.jpeg)

**Comment by pwoessner** [comment]
Thanks for the details! Very much appreciated. We are looking into the issue and trying to reproduce it. 

My current assumption:
In order to fix the error the app needs to update its data with the keys from the server.
There are 2 possibilities how this could be done: 

- Either trigger the update manually by deactivating the background activity in the app settings (which is not possible on all android os versions / brands). Please also note that you maybe have to wait until tomorrow because you can update the app only once a day and maybe the update was done today in the background
- Wait until tomorrow when the background job will be executed to update the app automatically

Another possibility could also be to reset the app data in the in-app settings which will bring the app in a consistent state immediately. Note: this will also delete any other data (e.g. tracing duration) that is stored in the app but **not** the google tracing information (contacts you had today and so on). 

Maybe to add some technical background (if you are interested):
This error is caused if the app has a date set when the app tried to update the random keys from the server but the data is not accessible in the risk calculation. Normally this state is impossible to reach because as soon as you fetch the keys from the server the exposure summary will be persisted. If that persistence fails the date of the update will also be deleted and therefore you only should see the "Unbekanntes Risiko" without any error.

**Comment by tvannahl** [comment]
I have to thank you for the great transparency and feedback and yeah I've been curious.

We'll wait until tomorrow and see if it's fixed by then.

**Comment by pwoessner** [comment]
FYI: The background job is happening at a random time during the day (7-23:59 if I am not mistaken) so it could take a while until the background job is being executed.

**Comment by hoehleatsu** [comment]
Experienced the same error. :-( I turn the phone into flight mode during nights (but bluetooth enabled) and it sounds like this causes the error, when it tries to fetch the keys?

"Da Sie die Risiko-Ermittlung noch nicht lange genug aktiviert haben, konnten wir für Sie kein Infektionsrisiko berechnen" as error message sounds innapropriate, because then I would like to know how many days would be needed and the problem seems to be elsewhere.

**Comment by corneliusroemer** [comment]
> Experienced the same error. :-( I turn the phone into flight mode during nights (but bluetooth enabled) and it sounds like this causes the error, when it tries to fetch the keys?
> 
> "Da Sie die Risiko-Ermittlung noch nicht lange genug aktiviert haben, konnten wir für Sie kein Infektionsrisiko berechnen" as error message sounds innapropriate, because then I would like to know how many days would be needed and the problem seems to be elsewhere.

@hoehleatsu You may want to report this as a new issue (if it doesn't yet exist). Wrong notification "nicht lange genug aktiviert" instead off "wasn't connected to the internet when trying to download exposure keys - thus cannot tell if exposed" when internet was off, blocking exposure key download.

**Comment by pgottinger** [comment]
Hi,
we have the same error (also same stack trace):
Device: Motorola Moto G7 Play
App installation date: 16th June
App version was 1.0.0, but just updated to 1.0.2 and error still occurs.

In case you need additional information, please let me know.

**Comment by pwoessner** [comment]
@tvannahl are you still experiencing the issue or is this now resolved for you?

**Comment by pwoessner** [comment]
Hi @hoehleatsu & @pgottinger , thanks for reporting the issue. 
Can you please share some more details? Am I assuming right that you also see the "Ursache 3500" error dialog? Or is your app "only" showing the Unkown Risk card?

Can you please also share some additional info how you used the app? (Internet on/off, Bluetooth on/off)

Thanks,
Philipp

**Comment by tvannahl** [comment]
> @tvannahl are you still experiencing the issue or is this now resolved for you?

Thank you, the issues appears to be fixed on our side. Until now it didn't show up again. Thank you for the support.

**Comment by hoehleatsu** [comment]
Thanks for replying and sorry for being unspecific: I experienced the exact same error message as @tvannahl, i.e. also the "Ursache 3500" with details containing "Ursache: x.a.a.f ....".

Usage of the App: Download on June 16, runs with bluetooth on the whole time, but with flight-mode enabled during nights (but ensuring that bluetooth is still on, despite of flight mode). Hence, no internet connection exists during the night.

After reading the various bug reports here, I've tried to see if the problem is connected to any type of battery optimization, but at the time of report I had not tried any configuration in this direction.

Not sure this makes my report clearer - otherwise let me know. I've now resetted the app and hope the next risk level calculations works the next time keys are fetched.

**Comment by pwoessner** [comment]
Hi all,

so we have most probably identified the issue. It will be fixed in #662. 

The issue was during the rollback of the diagnosis key download transaction. So if you happened to have one successful key retrieval transaction the date of the download and the token are set during that execution. The risk level calculation checks first the date (if you never downloaded keys then you will get the unknown risk because there is no way to know if you are potentially infected if you have not provided any diagnosis keys yet) and if you have a date set then we use the token to retrieve an exposure summary to check the exposures. 

In your case it happened that on the next day the second key retrieval failed (probably because of airplane mode). If the key retrieval transaction fails it has a rollback behaviour defined which will set all data back to its original value. Here it happened that the date was rollbacked to the date of the previous execution but due to a bug the token was set to null (and not to the token of the previous transaction). 

This caused the risk level calculation to jump over the risk unknown step because the key retrieval date is set but in the next step where it calculates the risk level it was unable to get the exposure summary because of the missing token.

As soon as you run through the next successful key retrieval this error should be fixed. 

We will include this fix in the next release.

Thanks for your help,
Philipp

**Comment by hoehleatsu** [comment]
Thanks for your quick response to this!

**Comment by jakobmoellersap** [comment]
The issue should be fixed in the latest release with 1.0.4
